Class/Tools.php

```diff
@@ -888,7 +1148,200 @@ public static function stripAttributes($html, $attributes)
                 $item->removeAttribute($attribute);
             }
         }
-        return substr($domd->saveXML(), 26, -6);
+        return substr($domd->saveXML(), 26, -6); //@todo: suboptimal, version-sensitive
+    }
+
+    /**
+     * Get GoogleAuthenticator hash of given secret and timeslot.
+     * https://en.wikipedia.org/wiki/Google_Authenticator
+     *
+     * @param string $secret 6-8 chars secret
+     * @param int $timeSlot delta to time slot of floor(time() / 30)
+     * @return int
+     */
+    public function GoogleAuthenticatorCode($secret, $timeSlot = 0)
+    {
+        if ($timeSlot < 1000) {
+            $timeSlot += floor(time() / 30);
+        }
+        $data = str_pad(pack('N', $timeSlot), 8, "\0", STR_PAD_LEFT);
+        $hash = hash_hmac('sha1', $data, $secret, true);
+        $unpacked = unpack('N', substr($hash, ord(substr($hash, -1)) & 0xF, 4));
+        return ($unpacked[1] & 0x7FFFFFFF) % 1000000;
+    }
+
+    /**
+     * Inverse to str_getcsv() - return comma-separated-values out of given array
+     * For parameter details see PHP's fputcsv()
+     * edited version from https://gist.github.com/johanmeiring/2894568
+     *
+     * @param array $fields
+     * @param string $delimiter (optional)
+     * @param string $enclosure (optional)
+     * @param string $escape_char (optional)
+     * @result string CSV of given arguments
+     */
+    public static function str_putcsv(array $fields, string $delimiter = ',', string $enclosure = '"', string $escape_char = "\\")
```

Comment:
PHPUnit (pro PHP/5.6) zde failuje s hláškou:
Fatal error: Default value for parameters with a class type hint can only be NULL in K:\Dropbox\Work\godsdev-github\tools\Class\Tools.php on line 1184
... string prostě ještě v PHP/5.6 nelze takto užít

Reply:
Díky - "string" ruším.